webapp/src/ts/services/transitions/create-user-for-contacts.transition.ts

```diff
@@ -0,0 +1,199 @@
+    super();
+  }
+
+  readonly name = 'create_user_for_contacts';
+  private readonly REPLACE_PROPERTY = 'replace_forms';
+  private replaceForms;
+
+  /**
+   * @param {Object} settings - the CHT instance settings
+   * @return {Boolean} - whether the required config is present
+   */
+  init(settings) {
+    if (!settings) {
+      return false;
+    }
+    this.replaceForms = this.getReplaceForms(settings);
+    if (!this.replaceForms || !Array.isArray(this.replaceForms) || !this.replaceForms.length) {
+      console.warn(
+        `Configuration error. Config must define have a '${this.name}.${this.REPLACE_PROPERTY}' array defined.`
+      );
+
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * @param {Array<Doc>} docs - docs to be saved
+   * @return {Boolean} - whether any of the docs from the batch should be processed
+   */
+  filter(docs: Doc[]) {
+    if (!docs) {
+      return false;
+    }
+    return !!docs.filter(doc => doc?.type === 'data_record').length;
+  }
+
+  /**
+   * @param {Array<Doc>} docs - docs to run the transition over
+   * @returns {Promise<Array<Doc>>} - updated docs (may include additional docs)
+   */
+  async run(docs: Doc[]) {
+    if (!docs) {
+      return [];
+    }
+    docs = docs.filter(doc => doc);
+
+    const originalContact = await this.userContactService.get({ hydrateLineage: false });
+    if (!originalContact) {
+      return docs;
+    }
+
+    if (this.createUserForContactsService.isBeingReplaced(originalContact)) {
+      this.reparentReports(docs, originalContact);
+    }
+    const userReplaceDoc = this.getUserReplaceDoc(docs);
+    if (userReplaceDoc) {
+      await this.replaceUser(docs, userReplaceDoc, originalContact);
+    }
+
+    return docs;
+  }
+
+  private getTransitionConfig(settings) {
+    return settings[this.name];
+  }
+
+  private getReplaceForms(settings) {
+    return this.getTransitionConfig(settings)?.[this.REPLACE_PROPERTY] || [];
+  }
+
+  private getUserReplaceDoc(docs: Doc[]) {
+    const replaceDocs = docs.filter(doc => this.replaceForms.includes(doc.form));
+    if (replaceDocs.length > 1) {
+      throw new Error('Only one user replace form is allowed to be submitted per transaction.');
+    }
+
+    return replaceDocs[0] as UserReplaceDoc;
+  }
+
+  private async replaceUser(docs: Doc[], userReplaceDoc: UserReplaceDoc, originalContact: Doc) {
+    const replacementContactId = userReplaceDoc.fields?.replacement_contact_id;
+    if (!replacementContactId) {
+      throw new Error('The form for replacing a user must include a replacement_contact_id field ' +
+        'containing the id of the new contact.');
+    }
+
+    const originalContactId = userReplaceDoc.contact?._id;
+
+    const hasOriginalContactId = originalContact._id === originalContactId;
+    const hasReplacedContactId = this.createUserForContactsService.isBeingReplaced(originalContact) &&
+      this.createUserForContactsService.getReplacedBy(originalContact) === originalContactId;
```

Comment:
This is here to support the scenario where user associated with contact `A` is replaced with a user associated with contact `B`, then before syncing another user replace form is submitted to replace with another new user associated with contact `C`.  Since no syncing has happened, the logged in user will still technically be associated with `A` and that will be the original `contact` set on the last user replace form when it is submitted. However, the re-parenting process in this transition will automatically update the `contact` on the form to be `B` as a part of this transition (since the user has been previously replaced).  

The problem is that `originalContact._id` in this function is still going to be `A` since that is the contact associated with the current user.  But, now `userReplaceDoc.contact._id` is `B` since it was re-parented.  So, to avoid throwing the following error in this case, we need to do this extra check to see if the contact associated with the user replace form matches.